gigascience/gigadb-website
# Gitlab  Runner cloud deployment and pipeline improvement

# Pull request for issue: #808

This is a pull request for the following functionalities:

Improve the speed of the deployment pipeline.

This is achieved with these changes:

## (1) Deploy Gitlab Runner on a cloud-based virtual machine to run the  build and deploy jobs of the pipeline

It needs to be running pipelines for all developers.
Elastic and auto-scalable setup for gitlab runners are complex and need much more time and investment to get right.
So a beefy host machine is the next best thing we can use. Because the developers are located across a wide timezone difference, 
the host machine will be in used almost continuously 24/7. That and because the runner deployment is quite self-contained and independent from the rest 
of our enterprise application infrastructure, it makes sense to use a paid monthly virtual machine like those from Amazon Lightsail, Digital Ocean, and Linode.
I'm familiar with DO, and our sftp server is running on droplet (DO's VM names), so it makes to reuse the same VM we use for the sftp server already.

However, after testing, the original set of capacity selected (1 GB RAM, 1 vCPU, 25 SSD, 1TB b/w) for the VM were not enough to run the Gitlab pipeline with good performance.
Increasing to (2 GB RAM, 2 vCPU, 25 SSD, 3TB b/w) gives good result when running my own pipeline. 
However I don't think after reading the usage stats when my pipeline run, that's enough for the three of us, especially that there would be at least two pipelines 
potentially running concurently in the Hong Kong time zone. 
So I've increased the capacity to (8 GB RAM, 4 vCPU, 25 SSD, 5TB b/w).

The gitlab runner run as Docker compose service and in the repository it has its own sub project as a GigaDB tool:

* gigadb/app/tools/gitlab-runner/


## (2) Reorganisation of pipeline stages to have acceptance tests after deployment to staging

The automated acceptance tests's purpose is business level validation of the implementation, so it's associated with, related to and complementary of the activities performed on the staging environment by the business, so it makes sense to have automated acceptance tests performed in a stage  (called ``Acceptance``) after deployment to staging has been performed. The side effect, is it shorten the time it takes the pipeline get to staging deployment, as only unit/functional/integration tests are performed before (and they need to stay that way as they validate the integrity  of the staging environment).

* .gitlab-ci.yml
* ops/pipelines/gigadb-test-jobs.yml

## (3) Migrate to PHP 7.3 to improve performance of PHP code

Although migrating to PHP 7.4 will be highly beneficial to both performances and developer productivity, there are high risks of incompatibility with our legacy codebase, especially the automated tests. 

 Migrating to PHP 7.3 was less painful and yielded some developer productivity (see [changes](https://www.zend.com/blog/guide-to-php-73)) and more importantly performance gains (I shaved 2mn of the build jobs after the migration). 

A couple of relatively minor changes had to be made to allow migration to PHP 7.3:

* Composer installation needed updated  to explicitly include a dependency (``libzip-dev``) previously unneeded. 
* The version of APCu needed to be upgraded to a version compatible with PHP 7.3

The APCu version was hard-coded, so I've refactored version managed for it and other dependencies with hard-coded version.

Details of this work can be seen in commits:

* rija/gigadb-website@211341b508e533d0cc5397d140374f66c89d8303
* rija/gigadb-website@de8309498f9fa58c23bf650f40c095e1f08b6fa8
* rija/gigadb-website@9c608b7a27996df3829ddd919b864babd36d00b8



## (4) Parallelise unit and functional tests on the pipeline

Parallelize unit/functional/integration tests

the dividing lines are:

* legacy tests (run with php units) and coverage calculation
* Codeception tests (unit and funcitonal)
* Integration tests (everything else)

The Codeception tests are independent of the legacy PHP Unit tests and run faster than them.
So do the integration tests. 
By having a dedicated pipeline job for each of them, they can run in parallel.

 the legacy tests which are the slowest run 1mn faster due to the removal of the the other two type of tests.

This organisation make it easy to deprecate the legacy tests in due time or to make structural changes to the other types of tests without impacting or without being block by the legacy tests.

Details for this work are in commit rija/gigadb-website@8d7022bf540fc4fc51c0ee6d3ef6710c11709ab0

## How to test

* Make sure the new variables defined in ``env-sample`` are in your .env with the same values
* Push this branch to your github origin so your fork can run a pipeline for it.

The first time pipeline run may be slow as your fork has no cache warmed up yet.
You can trigger the pipeline again by clicking "Run pipeline" and selecting your remote branch for this branch.
Subsequent runs should show the speed gain for the pipeline execution
